### ISSUE TITLE
Draft Ops policy for adding editor to Edit Button Dropdown

### ISSUE THREAD

**Issue by Firefishy** [openstreetmap/operations#877]
The OSM Operations team currently do not have a policy for adding an Editor to the openstreetmap.org edit dropdown menu.

We currently have policies for [Info Message](https://operations.osmfoundation.org/policies/banner/) and [Adding a new tile layer](https://operations.osmfoundation.org/policies/new-tile-layers/).

We should draft a policy for the requirements / expectations for adding an editor to the dropdown.

**Comment by Firefishy** [comment]
cc @grischard

**Comment by Firefishy** [comment]
For a laugh I got chat-gpt to draft a policy, it has some interesting bits: https://gist.github.com/Firefishy/a2aeb106a0a9736a84e654db7614dca5

**Comment by mvexel** [comment]
The ChatGPT version is actually a decent start 🤖 But does it make sense to have a separate policy for this? How many general-purpose editors exist or could we expect to be built that would be tested against it? The only one existing today that I can think of off the top of my head is Rapid. 

If we were to come up with a policy, would it make sense to have it be about links on osm.org to external tools more in general? For example on a changeset detail page you could link to OSMCha, Achavi etc, on a relation detail page to OSM Relation Analyzer, on the history page to Deep History. There's a lot of external tools that *could* be useful to link to for mappers. A policy to guide those decisions would help.

**Comment by tomhughes** [comment]
Well if you actually want a decision then I suggest not complicating things!

We have refused to add links to the sort of tools you mention on many occasions in the past and I don't really want to reopen that - the number of such tools which exist is one major reason why they create rather different issues.

One thing I'm still not entirely clear about with regard to the specific request that led to the creation of this policy is whether it was talking about linking to an external instance of the editor or integrating one in the way iD is - it's important because the things we need to consider are rather different for the two cases.

**Comment by mvexel** [comment]
> Well if you actually want a decision then I suggest not complicating things!

Our combined 30+ years experience in OSM would suggest that you are correct.

> One thing I'm still not entirely clear about with regard to the specific request that led to the creation of this policy is whether it was talking about linking to an external instance of the editor or integrating one in the way iD is - it's important because the things we need to consider are rather different for the two cases.

Not to veer off topic too much (since the ticket is about policy), integration in the same way iD is would be the goal.

**Comment by mvexel** [comment]
Specifically for Rapid (the integration I originally asked about): in the meantime I spoke with the Rapid team and they're happy to support with the actual integration work needed.

Back on the main topic, I agree that there needs to be some decision criteria for including new editors more generally. Taking a stab at criteria:

* General purpose / aimed at a general audience
* Desktop based and either web-based or cross-platform
* Open Source
* Actively supported and developed
* Community feedback actively welcomed
* Has validation of map edits which evolves with evolving tagging schema

This is just a start, but I would suggest to keep it fairly concise and as objective as possible to actually make it possible to come to a decision.

**Comment by mvexel** [comment]
I spent some time writing up [a draft](https://docs.google.com/document/d/1K4kf28iFJwOOeKJ9qYQBJQr-Im4yYzZW-IXDUMRL1i4/edit#), borrowing from and expanding on the already accepted tile layer inclusion policy. 

It would be great to discuss this in an upcoming OWG meeting. In the meantime the doc is open for comments.

**Comment by pnorman** [comment]
@Firefishy to send policy out for initial consultation

**Comment by Firefishy** [comment]
Published the draft for discussion: https://community.openstreetmap.org/t/proposal-osm-org-editor-inclusion-policy/106589

**Comment by eneerhut** [comment]
Exciting! Will this be a topic for the next OWG meeting?

**Comment by pnorman** [comment]
Changes decided on last meeting (May 2nd)
- drop reference the desktop editors because they already handled by the remote editor protocol policies about web based editors
- clarify that we are happy to add a generic mobile editor entry

A generic mobile editor entry would require mobile devs to come to an agreement on how this is best done

**Comment by eneerhut** [comment]
@pnorman if I understand correctly, a new post is going out to propose Rapid Editor as an additional option on openstreetmap.org. The post will be up for community discussion for 4 weeks. Is this correct?

Is this contingent on the changes above being made first?

**Comment by tordans** [comment]
> … to add a generic mobile editor entry
> A generic mobile editor entry would require mobile devs to come to an agreement on how this is best done

Can someone describe how this will work from a technical point of view? I am not too familiar with this area but AFAIK it will not be possible to just have one mobile edit button and have the apps or iOS decide which one will be opened. AFAIK – at least for iOS – we need one button per editor. 

I started this discussion to answer this question for GoMap in https://github.com/bryceco/GoMap/issues/758. There are a few links with more info in that post.

I don't see an issue from this policies point of view, though, to treat mobile editors like any other editor. We can decide to add some mobile editors per operating system. This would require the edit button to be visible for the right operating system only, which is something that can be added to the website code.

**Comment by bryceco** [comment]
For iOS you can use a generic URI like `geo:44.05404,-123.10062?z=18` which will open whatever app happens to be registered for geo: (maybe an editor, maybe a navigation app, etc). Notably, the user isn't provided a list of apps to choose from: the OS chooses for them, so it's incredibly frustrating in general.

Or you can name a specific website that deep-links to an associated app:
`https://gomaposm.com/edit?center=47.679056,-122.212559&zoom=21`
If the app is installed then the deep-link opens the app, and if it isn't installed it opens the website (for Go Map!! it just has a download button). 

I'm doubtful there's a mechanism that works equally well for both Android and iOS. IMO if we're special-casing desktop and mobile we might as well special-case desktop, Android and iOS. That way everyone gets the best possible experience.

**Comment by tomhughes** [comment]
> Can someone describe how this will work from a technical point of view? I am not too familiar with this area but AFAIK it will not be possible to just have one mobile edit button and have the apps or iOS decide which one will be opened. AFAIK – at least for iOS – we need one button per editor.

I don't know the technical details but I believe both iOS and Android have a way for apps to register to handle particular URLs or mime types or such so we would just need an agreed scheme they could register for? Certainly I often get Android ask what app I want to use to handle certain things.

Note that we don't really want to use `geo:` for this as we already have that in the share tab and viewers register for that as well as editors.

**Comment by pnorman** [comment]
`geo:` is also a poor choice because we have a bunch of parameters such as OSM ID, etc that we can end up passing.

**Comment by tordans** [comment]
Given the way iOS works as described in https://github.com/openstreetmap/operations/issues/877#issuecomment-2119116579 I think we should prepare this policy to allow each mobile editor that we agree to add to the website to be a separate entry in the list of mobile editors.

For iOS that will likely be GoMap!!, StreetComplete, EveryDoor.
For Android that will likely be Vespucci, StreetComplete, EveryDoor.

The edit URLs will then link to a domain defined by those editors which is where the editors register their corresponding apps and handle the case of when the App is not installed. (This is what some of them do right no now, already.)

Having such a domain which can handle certain variables like `osm_id`, `osm_type` _or_ `lat`, `log`, `zoom` to open the editor in a useful way and provides a fallback for when the app is not installed should IMO be a requirement for the editor to be considered.

**Comment by tomhughes** [comment]
My memory of the ops call is that our intended wording was that we want a single "mobile" entry if that was technically possible so I think that is already covered.

That said I really, really don't want an edit menu that has dozen entries on it...

**Comment by bryceco** [comment]
> we want a single "mobile" entry if that was technically possible 

It isn't technically possible to have a single entry for mobile. At least not for iOS. 

> I really, really don't want an edit menu that has dozen entries on it...

That shouldn't happen if we properly detect the platform and only show relevant editors, and if the policy limits the editors to only the most popular ones. We show multiple desktop editors when the user is on desktop, why wouldn't we show multiple mobile editors on mobile?

**Comment by simonpoole** [comment]
> The edit URLs will then link to a domain defined by those editors which is where the editors register their corresponding apps and handle the case of when the App is not installed. (This is what some of them do right no now, already.)

This is completely unnecessary on Android as you will get an app chooser if there are more than one installed app that can handle the intent. 

In particular, while the draft policy more or less says so, starting the app via the OSM website should do something useful, that is, allow editing of the currently displayed area even if you are currently not located there, and not simply be a link for marketing purposes.

**Comment by tordans** [comment]
@simonpoole we already [confirmed](https://github.com/openstreetmap/operations/issues/877#issuecomment-2130975303) that for iOS we need one entry per app. 

Given this, we could choose a different route for Android and have only one "Edit with mobile app" Entry which will trigger the app chooser which will present the user already installed apps.

I would rather see us use the same flow for both systems. But I am not familiar with the Android flow to know if the same setup would work. For iOS this is a domain which either triggers the app (like a redirect between apps) or shows some website (or maybe open the app store? (not sure)).

I would welcome the links in the menu to be a nudge for users to test out apps that we agreed on that are valuable to check out. This is also nothing new IMO, given the JOSM Links being there for me … even though I don't have JOSM installed.

**Comment by simonpoole** [comment]
To be clear the JOSM link is just a quite generic  URL / schema that can be implemented by essentially any app that would make sense to start from the menu in question (iOS issue noted), and historically has been supported by Merkaartor and Potlatch in its desktop incarnation too. I've suggested that the link it self be renamed to "Start editor with remove control" or something similar for a long time instead of currently referencing three editors.

And to just reiterate it is a tremendously bad idea to misuse the edit button as a marketing platform.

**Comment by AntonKhorev** [comment]
> "Start editor with remove control"

Remote control doesn't start an editor. It controls the editor that was already started.

**Comment by simonpoole** [comment]
On Android that is not so.

**Comment by AntonKhorev** [comment]
> On Android that is not so.

Only if you want to call custom uri handlers "remote control".

**Comment by simonpoole** [comment]
> Only if you want to call custom uri handlers "remote control".

That is a misunderstanding: the regular 127.0.0.1 http URL will work perfectly, including starting any app that has registered itself for such an intent, it is just that it wont work as source of an iframe.

**Comment by AntonKhorev** [comment]
So this is no longer true: https://vespucci.io/tutorials/vespucci_intents/#supported-intents ?

> Directly using the local host IP address with a port will not work on modern Android variants.

**Comment by simonpoole** [comment]
> So this is no longer true: https://vespucci.io/tutorials/vespucci_intents/#supported-intents ?
> 
> > Directly using the local host IP address with a port will not work on modern Android variants.

This was in reference to the iframe issue (the iframe aspect wasn't clear at the time), https://github.com/MarcusWolschon/osmeditor4android/blob/master/src/main/AndroidManifest.xml#L199 continues to exist and works in principle.

**Comment by AntonKhorev** [comment]
`iframe` is not the issue, the same is going to happen with ajax requests. If you do a http request explicitly, it's not going to get intercepted by an intent.

To make it work you'd have to replace the `href` of the edit button (which is actually a link styled as a button) and let a normal click event happen, similar to https://github.com/openstreetmap/openstreetmap-website/blob/948658b7ca8db5dac9ade64e32fb1fa89b03d62d/app/assets/javascripts/index.js#L287-L290. And I still wouldn't call it "remote control".

**Comment by AntonKhorev** [comment]
I tried an actual remote control from a phone by running josm on another pc and replacing `127.0.0.1` with its address, it worked. In theory we could have an option for remote control root url.

**Comment by AntonKhorev** [comment]
We're using `fetch` now https://github.com/openstreetmap/openstreetmap-website/pull/5375. If it launches Vespucci for anyone, please tell.

**Comment by jenniferlily236** [comment]
The OSM Operations team should create clear criteria for adding editors to the dropdown menu, similar to policies for tile layers and info messages. This should include stability, open-source availability, active maintenance, and community relevance. A transparent review process will ensure fairness and consistency—just as clear standards do in fields like [web design in Dubai](https://www.techigator.ae/services/web-design-company-dubai).
